### ISSUE TITLE
Should `AbstractLayout` and `AbstractTrace` be subtypes of `AbstractDict`?

### ISSUE THREAD

**Issue by thchr** [sglyon/PlotlyBase.jl#67]
I just found that e.g. `get!(::Layout, ...)` currently isn't implemented. 

The simplest way to add that (aside from just explicitly adding an implementation which is of course also possible) would be to make `AbstractLayout` and `AbstractTrace` subtypes of `AbstractDict`. Would there be any downside to this?

**Comment by thchr** [comment]
I think that e.g. `get!` should work immediately, cf. [the Base implementation](https://github.com/JuliaLang/julia/blob/6469ca538482d0751c69a3723ad97d2728da2f5e/base/abstractdict.jl#L509-L515), since you already have `haskey` and `getindex` implemented. I think the closest thing to a documented interface is the discussion in [this issue](https://github.com/JuliaLang/julia/issues/25941).

**Comment by sglyon** [comment]
that implementation looks good to me. Do you have a sense of what else we would pick up via AbstractDict?

**Comment by thchr** [comment]
Anything you already implemented specific methods for would stick around; e.g. `show` and so on. Any methods that were not previously explicitly overloaded would fall back to `AbstractDict`'s - i.e., it shouldn't change anything that you already are using, only enable behaviors beyond what's already implemented (but that is consistent with `AbstractDict` behavior).

One possible downside might be ambiguities: if there's some method like `f(::AbstractDict, ::ConcreteThing)` already defined in Base and you have `f(::Layout, ::AbstractThing)` overloaded here, the call would be ambiguous. Other than that, I cannot think of anything.

**Comment by thchr** [comment]
I may be misunderstanding the design choices for `AbstractLayout` and `AbstractTrace` - but isn't the intention that they behave exactly like dictionaries? Since you already have explicit methods defined for any "non-dictionary" behavior, what would be the downside of adopting dictionary behavior for anything "left out"?

Equivalently, I imagine you are not depending on any currently-undefined non-dict behavior of these types - so even if there's new behaviors added after subtyping it to `AbstractDict`, it seems like you wouldn't actually "hit" those behaviors.

**Comment by sglyon** [comment]
I think you have a great grasp of the design!

In general I want them to be similar to AbstractDict, but have some special treatment for setting/getting keys including underscores (to make things like `marker_line_color` easier to set)

Also, I agree that the code in this library wouldn't hit any currently-undefined behaviors. My concern is that some of the special treatment assumptions I made violate other assumptions inside base for how AbstractDict works. By subtyping AbstractDict I implicity commit to follow the associated contract/interface and I don't want to lie!

In the end I'm probably being to pedantic here and we would have net benefit by making the subtype explicit.  

tl;dr -- I'm willing to do it!
